File: assets/scripts/tunnel.clj
#!/usr/bin/env bb

(ns tunnel
  (:require
   [babashka.process :refer [process]]
   [babashka.cli :as cli]))

(def envs
  {:dev  {:local-port 0
          :db-host ""
          :db-port 0
          :bastion ""
          :user ""}})


(defn start-tunnel! [env]
  (let [{:keys [local-port db-host db-port bastion user]} (envs env)
        proc (process
              ["ssh"
               "-o" "ExitOnForwardFailure=yes"
               "-N"
               "-L" (str local-port ":" db-host ":" db-port)
               (str user "@" bastion)]
              {:inherit true})]
    (println "Tunnel started for" (name env))
    (println "PID:" (:pid proc))
    @proc)) ;; block so SSH keeps running


(defn -main [& args]
  (let [{:keys [env]} (cli/parse-opts args)
        env (keyword env)]
    (when-not (contains? envs env)
      (throw (ex-info "Unknown env"
                      {:env env
                       :available (keys envs)})))
    (start-tunnel! env)))

(apply -main *command-line-args*)

File: assets/scripts/tunnel.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'optparse'

ENVS = {
  dev: {
    local_port: 0,
    db_host:    "",
    db_port:    0,
    bastion:    "",
    user:       "",
  },
}.freeze

def start_tunnel!(env)
  cfg = ENVS.fetch(env) do
    abort("!! Unknown env: #{env} — available: #{ENVS.keys.join(', ')}")
  end
  cmd = [
    "ssh",
    "-o", "ExitOnForwardFailure=yes",
    "-N",
    "-L", "#{cfg[:local_port]}:#{cfg[:db_host]}:#{cfg[:db_port]}",
    "#{cfg[:user]}@#{cfg[:bastion]}",
  ]
  puts "Tunnel started for #{env} (PID: #{Process.pid})"
  exec(*cmd)
end

def parse_opts(argv)
  opts = {}
  OptionParser.new do |o|
    o.on("-e", "--env ENV", "Environment to tunnel to") { |v| opts[:env] = v.to_sym }
  end.parse(argv)
  opts
end

def main(argv)
  opts = parse_opts(argv)
  abort("!! --env required — available: #{ENVS.keys.join(', ')}") unless opts[:env]
  start_tunnel!(opts[:env])
end

main(ARGV)
